src/devices_models/devices/common/cost_functions.jl

```diff
@@ -69,118 +76,132 @@ end
 Add variables to the OptimizationContainer for a service.
 """
 function cost_function!(
-    optimization_container::OptimizationContainer,
+    container::OptimizationContainer,
     devices::IS.FlattenIteratorWrapper{T},
-    ::DeviceModel{T, U},
+    model::DeviceModel{T, U},
     ::Type{<:PM.AbstractPowerModel},
-    feedforward::Union{Nothing, AbstractAffectFeedForward} = nothing,
 ) where {T <: PSY.Component, U <: AbstractDeviceFormulation}
     for d in devices
-        spec = AddCostSpec(T, U, optimization_container)
-        @debug T, spec
-        services = PSY.get_services(d)
-        add_service_variables!(spec, services)
-        add_to_cost!(optimization_container, spec, PSY.get_operation_cost(d), d)
+        spec = AddCostSpec(T, U, container)
+        @debug T, spec _group = LOG_GROUP_COST_FUNCTIONS
+        service_models = get_services(model)
+        add_service_variables!(spec, service_models)
+        add_to_cost!(container, spec, PSY.get_operation_cost(d), d)
     end
     return
 end
 
 function add_to_cost_expression!(
-    optimization_container::OptimizationContainer,
+    container::OptimizationContainer,
     cost_expression::JuMP.AbstractJuMPScalar,
-)
+    component::T,
+    time_period::Int,
+) where {T <: PSY.Component}
     T_ce = typeof(cost_expression)
-    T_cf = typeof(optimization_container.cost_function)
+    T_cf = typeof(container.cost_function)
     if T_cf <: JuMP.GenericAffExpr && T_ce <: JuMP.GenericQuadExpr
-        optimization_container.cost_function += cost_expression
+        container.cost_function += cost_expression
     else
-        JuMP.add_to_expression!(optimization_container.cost_function, cost_expression)
+        JuMP.add_to_expression!(container.cost_function, cost_expression)
+    end
+    if has_expression(container, ProductionCostExpression, T)
```

Comment:
@claytonpbarrows here we add the cost expressions

Reply:
@josephmckinsey , this is the new feature that will allow us to make an ex-poste calculation of generator costs. We'll need to make an extension for emissions to do the same thing for emissions costs